src/vs/workbench/contrib/positronNotebook/browser/contrib/undoRedo/positronNotebookUndoRedo.ts

```diff
@@ -0,0 +1,108 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (C) 2025 Posit Software, PBC. All rights reserved.
+ *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Disposable } from '../../../../../../base/common/lifecycle.js';
+import { WorkbenchPhase, registerWorkbenchContribution2 } from '../../../../../common/contributions.js';
+import { IEditorService } from '../../../../../services/editor/common/editorService.js';
+import { UndoCommand, RedoCommand } from '../../../../../../editor/browser/editorExtensions.js';
+import { IPositronNotebookService } from '../../../../../services/positronNotebook/browser/positronNotebookService.js';
+import { PositronNotebookEditorInput } from '../../PositronNotebookEditorInput.js';
+import { POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED, POSITRON_NOTEBOOK_CELL_EDITOR_FOCUSED } from '../../../../../services/positronNotebook/browser/ContextKeysManager.js';
+import { IUndoRedoService } from '../../../../../../platform/undoRedo/common/undoRedo.js';
+import { IContextKeyService } from '../../../../../../platform/contextkey/common/contextkey.js';
+
+class PositronNotebookUndoRedoContribution extends Disposable {
+
+	static readonly ID = 'workbench.contrib.positronNotebookUndoRedo';
+
+	constructor(
+		@IEditorService private readonly editorService: IEditorService,
+		@IUndoRedoService private readonly undoRedoService: IUndoRedoService,
+		@IPositronNotebookService private readonly positronNotebookService: IPositronNotebookService,
+		@IContextKeyService private readonly contextKeyService: IContextKeyService
+	) {
+		super();
+
+		const PRIORITY = 105;
+
+		this._register(UndoCommand.addImplementation(PRIORITY, 'positron-notebook-undo-redo', () => this.handleUndo()));
+		this._register(RedoCommand.addImplementation(PRIORITY, 'positron-notebook-undo-redo', () => this.handleRedo()));
+	}
+
+	private shouldHandleUndoRedo(): boolean {
+		const activeEditor = this.editorService.activeEditor;
+		if (!(activeEditor instanceof PositronNotebookEditorInput)) {
+			return false;
+		}
+
+		// Get the active notebook instance to access its scoped context key service
+		const instance = this.positronNotebookService.getActiveInstance();
```

Comment:
Do we need both the `activeEditor` check and `getActiveInstance`?

Reply:
Good catch, we do not. Fixed